### PR TITLE
 PreciseAnswer feature support

### LINKED ISSUE #2523: [QnAMaker} Sample update for Precise Answer Feature Request.
Update Samples for Precise Answer changes.

### PR DESCRIPTION
**Pull request by vipeketi**
Fixes #2523 
## Proposed Changes
Appsettings Request and Response changes for integrating Precise Answer changes in SDK. for C#

  -

### CODE REVIEW

File: samples/csharp_dotnetcore/49.qnamaker-all-features/BotServices.cs [hunk 1]
```diff
@@ -16,9 +16,16 @@ public BotServices(IConfiguration configuration)
                 EndpointKey = configuration["QnAEndpointKey"],
                 Host = GetHostname(configuration["QnAEndpointHostName"])
             });
+
+            this.EnablePreciseAnswer = bool.Parse(configuration["EnablePreciseAnswer"]);
+            this.ContentChoice = configuration["ContentChoice "];
```
Comment (somigithub) [inline samples/csharp_dotnetcore/49.qnamaker-all-features/BotServices.cs:21]: PreciseAnswerDisplayType : Single, Both

File: samples/csharp_dotnetcore/49.qnamaker-all-features/Dialog/QnAMakerBaseDialog.cs [hunk 2]
```diff
@@ -43,11 +43,13 @@ protected override Task<QnAMakerOptions> GetQnAMakerOptionsAsync(DialogContext d
             return Task.FromResult(new QnAMakerOptions
             {
                 ScoreThreshold = DefaultThreshold,
-                Top = DefaultTopN,
+                Top = 3,
```
Comment (somigithub) [inline samples/csharp_dotnetcore/49.qnamaker-all-features/Dialog/QnAMakerBaseDialog.cs:46]: revert

File: samples/csharp_dotnetcore/49.qnamaker-all-features/Dialog/QnAMakerBaseDialog.cs [hunk 3]
```diff
@@ -40,14 +45,17 @@ protected async override Task<IQnAMakerClient> GetQnAMakerClientAsync(DialogCont
 
         protected override Task<QnAMakerOptions> GetQnAMakerOptionsAsync(DialogContext dc)
         {
+
```
Comment (somigithub) [inline samples/csharp_dotnetcore/49.qnamaker-all-features/Dialog/QnAMakerBaseDialog.cs:48]: extra line

File: samples/csharp_dotnetcore/49.qnamaker-all-features/Dialog/QnAMakerBaseDialog.cs [hunk 4]
```diff
@@ -40,14 +45,17 @@ protected async override Task<IQnAMakerClient> GetQnAMakerClientAsync(DialogCont
...
             {
                 ScoreThreshold = DefaultThreshold,
                 Top = DefaultTopN,
                 QnAId = 0,
                 RankerType = "Default",
-                IsTest = false
-            });
+                IsTest = false,
+                EnablePreciseAnswer = this.EnablePreciseAnser
+
```
Comment (somigithub) [inline samples/csharp_dotnetcore/49.qnamaker-all-features/Dialog/QnAMakerBaseDialog.cs:57]: extra line, remove

File: samples/csharp_dotnetcore/49.qnamaker-all-features/Dialog/QnAMakerBaseDialog.cs [hunk 5]
```diff
@@ -63,9 +71,52 @@ protected async override Task<QnADialogResponseOptions> GetQnAResponseOptionsAsy
...
+                    }
+                }
+                else
+                {
+                    return false;
+                }
+            }
+        }
+
+        //DisplayPreciseAnswerOnly
```
Comment (somigithub) [inline samples/csharp_dotnetcore/49.qnamaker-all-features/Dialog/QnAMakerBaseDialog.cs:104]: remove this comment and extra line below

File: samples/csharp_dotnetcore/49.qnamaker-all-features/IBotServices.cs [hunk 6]
```diff
@@ -7,6 +7,6 @@ namespace Microsoft.BotBuilderSamples
 {
     public interface IBotServices
     {
-        QnAMaker QnAMakerService { get; }
+        QnAMaker QnAMakerService { get; }
```
Comment (somigithub) [inline samples/csharp_dotnetcore/49.qnamaker-all-features/IBotServices.cs:10]: revert this file change

File: samples/csharp_dotnetcore/49.qnamaker-all-features/appsettings.json [hunk 7]
```diff
@@ -1,7 +1,10 @@
...
   "MicrosoftAppId": "",
   "MicrosoftAppPassword": "",
-  "QnAKnowledgebaseId": "",
-  "QnAEndpointKey": "",
-  "QnAEndpointHostName": ""
+  "QnAKnowledgebaseId": "135099bb-f440-4832-be2b-9570bd530215",
+  "QnAEndpointKey": "670bd815e29b4e5ba4e4dccdfb8dd21e",
+  "QnAEndpointHostName": "https://qna-v2-mrc-demo.cognitiveservices.azure.com/qnamaker/v5.0-preview.1",
+  "QnAServiceType": "v2",
+  "EnablePreciseAnswer": "false",
```
Comment (somigithub) [inline samples/csharp_dotnetcore/49.qnamaker-all-features/appsettings.json:8]: make to check-in this file with 3 additional fields added + empty values for all

File: samples/csharp_dotnetcore/49.qnamaker-all-features/BotServices.cs [hunk 8]
```diff
@@ -14,8 +14,8 @@ public BotServices(IConfiguration configuration)
             {
                 KnowledgeBaseId = configuration["QnAKnowledgebaseId"],
                 EndpointKey = configuration["QnAEndpointKey"],
-                Host = GetHostname(configuration["QnAEndpointHostName"])
-            });
+                Host = GetHostname(configuration["QnAEndpointHostName"]),
```
Comment (somigithub) [inline samples/csharp_dotnetcore/49.qnamaker-all-features/BotServices.cs:17]: revert this file, it shouldn't be in this PR

File: samples/csharp_dotnetcore/49.qnamaker-all-features/Dialog/QnAMakerBaseDialog.cs [hunk 9]
```diff
@@ -63,9 +67,49 @@ protected async override Task<QnADialogResponseOptions> GetQnAResponseOptionsAsy
                 CardNoMatchText = DefaultCardNoMatchText,
                 NoAnswer = noAnswer,
                 CardNoMatchResponse = cardNoMatchResponse,
+                DisplayPreciseAnswerOnly = this.DisplayPreciseAnswerOnly
             };
 
             return responseOptions;
         }
+        private bool EnablePreciseAnser
```
Comment (cleemullins) [inline samples/csharp_dotnetcore/49.qnamaker-all-features/Dialog/QnAMakerBaseDialog.cs:75]: spelling

File: samples/csharp_dotnetcore/49.qnamaker-all-features/Dialog/QnAMakerBaseDialog.cs [hunk 10]
```diff
@@ -63,9 +67,49 @@ protected async override Task<QnADialogResponseOptions> GetQnAResponseOptionsAsy
...
+        {
+            get
+            {
+                var qnaServiceType = _config["QnAServiceType"];
+                if (string.Equals("v2",qnaServiceType, StringComparison.OrdinalIgnoreCase))
```
Comment (cleemullins) [inline samples/csharp_dotnetcore/49.qnamaker-all-features/Dialog/QnAMakerBaseDialog.cs:80]: Complicated and fragile. Please simplify.

File: samples/csharp_dotnetcore/49.qnamaker-all-features/Dialog/QnAMakerBaseDialog.cs [hunk 11]
```diff
@@ -20,17 +22,18 @@ public class QnAMakerBaseDialog : QnAMakerDialog
...
         public const string DefaultCardNoMatchResponse = "Thanks for the feedback.";
-
         private readonly IBotServices _services;
+        private readonly IConfiguration _config;
 
         /// <summary>
         /// Initializes a new instance of the <see cref="QnAMakerBaseDialog"/> class.
         /// Dialog helper to generate dialogs.
         /// </summary>
         /// <param name="services">Bot Services.</param>
```
Comment (mrivera-ms) [inline samples/csharp_dotnetcore/49.qnamaker-all-features/Dialog/QnAMakerBaseDialog.cs:32]: >  /// <param name="services">Bot Services.</param> [](start = 6, length = 50)

Add new param

**Comment by mrivera-ms** [comment]
>         }

There is no 'await' in this method.

---
Refers to: samples/csharp_dotnetcore/49.qnamaker-all-features/Dialog/QnAMakerBaseDialog.cs:42 in 9701d85. [](commit_id = 9701d8569c0e5210ae9bdf75921404fce3b6b16e, deletion_comment = False)

File: samples/csharp_dotnetcore/49.qnamaker-all-features/Dialog/QnAMakerBaseDialog.cs (hunk 9 above)
Comment (mrivera-ms) [inline samples/csharp_dotnetcore/49.qnamaker-all-features/Dialog/QnAMakerBaseDialog.cs:75]: nit: Add line break between methods.

File: samples/csharp_dotnetcore/49.qnamaker-all-features/Dialog/QnAMakerBaseDialog.cs [hunk 12]
```diff
@@ -63,9 +67,49 @@ protected async override Task<QnADialogResponseOptions> GetQnAResponseOptionsAsy
...
+            {
+                var qnaServiceType = _config["QnAServiceType"];
+                if (string.Equals("v2",qnaServiceType, StringComparison.OrdinalIgnoreCase))
+                {
+                    var rawEnablePreciseAnswer = _config["EnablePreciseAnswer"];
+                    if (!string.IsNullOrWhiteSpace(rawEnablePreciseAnswer))
+                    {
+                        return bool.Parse(rawEnablePreciseAnswer);
+                    }
+                    else
```
Comment (mrivera-ms) [inline samples/csharp_dotnetcore/49.qnamaker-all-features/Dialog/QnAMakerBaseDialog.cs:87]: >    else [](start = 16, length = 8)

Not needed.

File: samples/csharp_dotnetcore/49.qnamaker-all-features/Dialog/QnAMakerBaseDialog.cs [hunk 13]
```diff
@@ -63,9 +67,49 @@ protected async override Task<QnADialogResponseOptions> GetQnAResponseOptionsAsy
...
+                    {
+                        return true;
+                    }
+                }
+                else
```
Comment (mrivera-ms) [inline samples/csharp_dotnetcore/49.qnamaker-all-features/Dialog/QnAMakerBaseDialog.cs:92]: >else [](start = 16, length = 4)

Not needed.

File: samples/csharp_dotnetcore/49.qnamaker-all-features/Dialog/QnAMakerBaseDialog.cs [hunk 14]
```diff
@@ -63,9 +67,49 @@ protected async override Task<QnADialogResponseOptions> GetQnAResponseOptionsAsy
...
+        private bool DisplayPreciseAnswerOnly
+        {
+            get
+            {               
+                var rawDisplayPreciseAnswerOnly = _config["DisplayPreciseAnswerOnly"];
+                if (!string.IsNullOrWhiteSpace(rawDisplayPreciseAnswerOnly))
+                {
+                    return bool.Parse(rawDisplayPreciseAnswerOnly);
+                }
+                else
```
Comment (mrivera-ms) [inline samples/csharp_dotnetcore/49.qnamaker-all-features/Dialog/QnAMakerBaseDialog.cs:108]: >else [](start = 16, length = 4)

Redundant.

File: samples/csharp_dotnetcore/49.qnamaker-all-features/Dialog/RootDialog.cs [hunk 15]
```diff
@@ -22,10 +23,10 @@ public class RootDialog : ComponentDialog
         /// Initializes a new instance of the <see cref="RootDialog"/> class.
         /// </summary>
         /// <param name="services">Bot Services.</param>
```
Comment (mrivera-ms) [inline samples/csharp_dotnetcore/49.qnamaker-all-features/Dialog/RootDialog.cs:25]: >/// <param name="services">Bot Services.</param> [](start = 8, length = 48)

Add config param

File: samples/csharp_dotnetcore/49.qnamaker-all-features/Dialog/QnAMakerBaseDialog.cs (hunk 14 above)
Comment (vipeketi) [reply samples/csharp_dotnetcore/49.qnamaker-all-features/Dialog/QnAMakerBaseDialog.cs:112]: Implemented review comment.

---
In reply to: [433456660](https://github.com/microsoft/BotBuilder-Samples/pull/2430#discussion_r433456660) [](ancestors = 433456660)

File: samples/csharp_dotnetcore/49.qnamaker-all-features/Dialog/QnAMakerBaseDialog.cs (hunk 11 above)
Comment (vipeketi) [reply samples/csharp_dotnetcore/49.qnamaker-all-features/Dialog/QnAMakerBaseDialog.cs:33]: Added.

---
In reply to: [433454852](https://github.com/microsoft/BotBuilder-Samples/pull/2430#discussion_r433454852) [](ancestors = 433454852)

File: samples/csharp_dotnetcore/49.qnamaker-all-features/Dialog/QnAMakerBaseDialog.cs (hunk 9 above)
Comment (vipeketi) [reply samples/csharp_dotnetcore/49.qnamaker-all-features/Dialog/QnAMakerBaseDialog.cs:79]: Modified.

---
In reply to: [430667510](https://github.com/microsoft/BotBuilder-Samples/pull/2430#discussion_r430667510) [](ancestors = 430667510)

**Comment by somigithub** [comment]
reviewing in progress

File: samples/csharp_dotnetcore/49.qnamaker-all-features/appsettings.json [hunk 16]
```diff
@@ -3,5 +3,8 @@
   "MicrosoftAppPassword": "",
   "QnAKnowledgebaseId": "",
   "QnAEndpointKey": "",
-  "QnAEndpointHostName": ""
-}
\ No newline at end of file
+  "QnAEndpointHostName": "",
```
Comment (somigithub) [inline samples/csharp_dotnetcore/49.qnamaker-all-features/appsettings.json:6]: in review

File: samples/csharp_dotnetcore/49.qnamaker-all-features/QnABotAllFeatures.csproj [hunk 17]
```diff
@@ -8,9 +8,9 @@
 
   <ItemGroup>
     <PackageReference Include="Microsoft.AspNetCore.Mvc.NewtonsoftJson" Version="3.1.1" />
-    <PackageReference Include="Microsoft.Bot.Builder.AI.QnA" Version="4.9.2" />
-    <PackageReference Include="Microsoft.Bot.Builder.Dialogs" Version="4.9.2" />
-    <PackageReference Include="Microsoft.Bot.Builder.Integration.AspNet.Core" Version="4.9.2" />
+    <PackageReference Include="Microsoft.Bot.Builder.AI.QnA" Version="4.10.0-preview-200621-139841" />
+    <PackageReference Include="Microsoft.Bot.Builder.Dialogs" Version="4.10.0-preview-200621-139841" />
+    <PackageReference Include="Microsoft.Bot.Builder.Integration.AspNet.Core" Version="4.10.0-preview-200621-139841" />
```
Comment (cleemullins) [inline samples/csharp_dotnetcore/49.qnamaker-all-features/QnABotAllFeatures.csproj:13]: Production samples can't depend on packages that are not on Nuget. These packages are from our daily feed, which won't work here. This effectivly blocks this sample from merging.

Comment (somigithub) [reply samples/csharp_dotnetcore/49.qnamaker-all-features/QnABotAllFeatures.csproj:13]: @vipeketi 

this PR no longer required with new preview feature approach.

please pull it back.

**Pull request closed without merging**